### PR TITLE
Avoid adding connection to marked deleted elements upon delete in-place in tiered index [MOD-7732]

### PR DESCRIPTION
**Pull request by alonre24**
**Describe the changes in the pull request**

Upon in-place deletion of vectors that occur after asynchronous deletion in the tiered index, we did not properly handle the repairs for the "marked deleted" vectors, as we allowed updating new connections to such elements. These elements will not be accounted for in the asynchronous repair for the marked deleted vector and may cause corruption in the graph state along the way. 
This PR fixes this by checking the elements selected upon repairing a node due to in-place delete and NOT updating new connections to deleted vectors.   

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 1]
```diff
@@ -1642,9 +1664,16 @@ void HNSWIndex<DataType, DistType>::removeAndSwap(idType internalId) {
...
+            // Note that in case of in-place delete, we might have not accounted for this edge in
+            // in the undirectional edges, since there is no point in keeping it there temporarily
+            // (we know we will get here and remove this deleted id permanently).
+            // However, upon asynchronoush delete, this should always succeed since we do update
+            // the incoming edges in the mutual update even for deleted elements.
+            bool res = neighbour.removeIncomingUnidirectionalEdgeIfExists(internalId);
+            // Assert the logical condition of: is_marked_deleted(id) => res==True.
+            (void)res;
+            // assert((!isMarkedDeleted(internalId) || res) && "The edge should be in the incoming
+            // unidirectional edges");
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:1676]: why is the assert commented out?

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 2]
```diff
@@ -1007,24 +1018,35 @@ void HNSWIndex<DataType, DistType>::repairConnectionsForDeletion(
...
+            // We will add the selected element as a new neighbor ONLY IF:
+            // 1. The selected element was already a connection before - hence, no update performed.
+            // 2. Both the neighbour which we are updating and the selected element to connect are
+            // not deleted, hence we do not create new connections for marked deleted elements.
+            if (neighbour_orig_neighbours_set[chosen_cand] ||
+                (!isMarkedDeleted(neighbour_id) && !isMarkedDeleted(chosen_cand))) {
+                neighbor_level.setLinkAtPos(new_num_links++, chosen_cand);
+            }
+        }
+        neighbor_level.setNumLinks(new_num_links);
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:1032]: aren't we locked here when performing this change? Do we have to check the deletion mark on every iteration?
Also the comment above should be fixed

File: src/VecSim/algorithms/hnsw/hnsw.h [hunk 3]
```diff
@@ -988,7 +988,18 @@ void HNSWIndex<DataType, DistType>::repairConnectionsForDeletion(
...
+            // We will add the selected element as a new neighbor ONLY IF:
+            // 1. The selected element was already a connection before - hence, no update performed.
+            // 2. Both the neighbour which we are updating and the selected element to connect are
+            // not deleted, hence we do not create new connections for marked deleted elements.
+            if (neighbour_orig_neighbours_set[chosen_cand.second] ||
+                (!isMarkedDeleted(neighbour_id) && !isMarkedDeleted(chosen_cand.second))) {
+                neighbor_level.setLinkAtPos(new_num_links++, chosen_cand.second);
+            }
+        }
+        neighbor_level.setNumLinks(new_num_links);
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw.h:1002]: aren't we locked here when performing this change? Do we have to check the deletion mark on every iteration?

File: src/VecSim/algorithms/hnsw/hnsw.h (hunk 1 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw.h:1676]: oops. uncommenting

**Review by GuyAv46 — DISMISSED** [review]

File: src/VecSim/algorithms/hnsw/graph_data.h [hunk 4]
```diff
@@ -58,6 +58,17 @@ struct ElementLevelData {
...
     void appendLink(idType node_id) { this->links[this->numLinks++] = node_id; }
+    void removeLink(idType node_id) {
+        size_t i = 0;
+        for (; i < numLinks; i++) {
+            if (links[i] == node_id) {
+                links[i] = links[numLinks - 1];
+                break;
+            }
+        }
+        assert(i < numLinks); // node_id not found - curroption in graph
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/graph_data.h:69]: The failing assert will appear in the log - added a string to the assert

**Review by meiravgri — APPROVED** [review]

**Pull request merged**